# Week 5 — Day 1: Meet Pandas (Series & DataFrame)
### Code to AI — Python & Data Science (from Zero to Portfolio)
**Instructor:** Ali Hasnain — Lead AI Engineer & CTO

---

## What you will learn today
1. Why Pandas exists, and how it is different from NumPy
2. The **Series** — a labeled 1D column of data
3. The **DataFrame** — a full table with rows and columns
4. Loading a **real dataset** (Titanic)
5. The 5-command ritual: how to look at *any* new dataset
6. Selecting columns
7. Quick summaries: mean, max, value_counts, unique

**Session plan (75 minutes)**

| Time | Section |
|------|---------|
| 0:00 – 0:05 | Recap of Week 4 + why Pandas |
| 0:05 – 0:20 | Series |
| 0:20 – 0:35 | DataFrame |
| 0:35 – 0:45 | Loading real data (Titanic) |
| 0:45 – 0:60 | The 5-command ritual |
| 0:60 – 0:70 | Selecting columns + quick summaries |
| 0:70 – 0:75 | Wrap-up + assignment |


## 1. Why Pandas?

In Week 4 we used **NumPy**. NumPy arrays are fast, but they have one big limit:
**everything is a number, and columns have no names.**

Real data is not like that. Real data looks like a spreadsheet:

| name | age | city | fare |
|------|-----|------|------|
| Ali | 25 | Karachi | 7.25 |
| Sara | 31 | Lahore | 71.28 |

Different types in different columns. Every column has a **name**.

**Pandas** gives us exactly this: a table in Python.

- NumPy = a calculator for numbers
- Pandas = a spreadsheet you can program

Pandas is the tool data scientists spend most of their day inside. From today until the end of this course, it is our main tool.

In [2]:
!pip install pandas -q

In [3]:
# The standard import. Everyone in the world writes it exactly like this.
import pandas as pd
import numpy as np

print("Pandas version:", pd.__version__)

Pandas version: 3.0.3


## 2. The Series — a labeled 1D column

A **Series** is one column of data.
It has two parts:
- the **values** (the data itself)
- the **index** (the label of each value)

Think of it as a list that remembers names.

In [5]:
list1 = [1,2,3,4,5]
print(list1)

[1, 2, 3, 4, 5]


In [4]:
# Creating a Series from a list
marks = pd.Series([85, 92, 78, 95])
print(marks)

0    85
1    92
2    78
3    95
dtype: int64


Look at the output carefully.

The left column `0 1 2 3` is the **index** (automatic, like list positions).
The right column is the **values**.
`dtype: int64` tells us the data type.

Now let's give the index our own labels.

In [6]:
marks = pd.Series([85, 92, 78, 95], index=["Ali", "Sara", "Bilal", "Ayesha"])
print(marks)

Ali       85
Sara      92
Bilal     78
Ayesha    95
dtype: int64


In [7]:
# The three things every Series has
print("Values:", marks.values)
print("Index :", marks.index)
print("Dtype :", marks.dtype)

Values: [85 92 78 95]
Index : Index(['Ali', 'Sara', 'Bilal', 'Ayesha'], dtype='str')
Dtype : int64


### Creating a Series from a dictionary

A dictionary already has keys and values — which is exactly index and values.
Pandas understands this automatically.

In [8]:
marks = pd.Series({"Ali": 85, "Sara": 92, "Bilal": 78, "Ayesha": 95})
print(marks)

Ali       85
Sara      92
Bilal     78
Ayesha    95
dtype: int64


### Indexing a Series: by label and by position

In [9]:
# By label
print("Sara's marks:", marks["Sara"])



Sara's marks: 92


In [12]:
# By position (like a list)
print("First student's marks:", marks.iloc[0])
print("First student's marks:", marks.iloc[1])
print("First student's marks:", marks.iloc[0:2])


First student's marks: 85
First student's marks: 92
First student's marks: Ali     85
Sara    92
dtype: int64


In [14]:
# Multiple labels at once
marks_2 = marks[["Ali", "Bilal"]]
print(marks_2)
print(marks[["Ali", "Ayesha"]])

Ali      85
Bilal    78
dtype: int64
Ali       85
Ayesha    95
dtype: int64


In [16]:
marks

Ali       85
Sara      92
Bilal     78
Ayesha    95
dtype: int64

In [17]:
# Math works on the whole Series at once, like NumPy
print(marks + 5)

Ali        90
Sara       97
Bilal      83
Ayesha    100
dtype: int64


## 3. The DataFrame — rows and columns

A **DataFrame** is a full table.

The easiest mental model:
> A DataFrame is a dictionary of Series. Each column is a Series. They all share the same index.

Let's build one from a dictionary of lists.

In [22]:
students = pd.DataFrame({
    "name":  ["Ali", "Sara", "Bilal", "Ayesha", "Hassan", "Asfand"],
    "age":   [25, 31, 22, 28, 35, 22],
    "city":  ["Karachi", "Lahore", "Karachi", "Islamabad", "Lahore", "Umerkot"],
    "marks": [85, 92, 78, 95, 66, 95]
})

print(students)

     name  age       city  marks
0     Ali   25    Karachi     85
1    Sara   31     Lahore     92
2   Bilal   22    Karachi     78
3  Ayesha   28  Islamabad     95
4  Hassan   35     Lahore     66
5  Asfand   22    Umerkot     95


In [23]:
students

,name,age,city,marks
0,Ali,25,Karachi,85
1,Sara,31,Lahore,92
2,Bilal,22,Karachi,78
3,Ayesha,28,Islamabad,95
4,Hassan,35,Lahore,66
5,Asfand,22,Umerkot,95


Notice we wrote just `students` at the end of the cell instead of `print(students)`.
In a notebook, this shows a nicely formatted table. Use it for DataFrames.

### The shape of a DataFrame

In [24]:
students.shape

(6, 4)

In [25]:
r, c = students.shape
print(f"The dataset contain {r} rows and {c} columns")

The dataset contain 6 rows and 4 columns


In [26]:
students.columns

Index(['name', 'age', 'city', 'marks'], dtype='str')

In [27]:
students.index

RangeIndex(start=0, stop=6, step=1)

In [28]:
students.dtypes

name       str
age      int64
city       str
marks    int64
dtype: object

In [29]:
print("Shape (rows, columns):", students.shape)
print("Columns:", students.columns)
print("Index  :", students.index)
print()
print("Data types of each column:")
print(students.dtypes)

Shape (rows, columns): (6, 4)
Columns: Index(['name', 'age', 'city', 'marks'], dtype='str')
Index  : RangeIndex(start=0, stop=6, step=1)

Data types of each column:
name       str
age      int64
city       str
marks    int64
dtype: object


In [30]:
students.info()

<class 'pandas.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   name    6 non-null      str  
 1   age     6 non-null      int64
 2   city    6 non-null      str  
 3   marks   6 non-null      int64
dtypes: int64(2), str(2)
memory usage: 324.0 bytes


In [ ]:
# .info() gives a full summary in one command
students.info()

`.info()` tells us three important things:
1. how many rows
2. the name and type of every column
3. how many **non-null** (not missing) values each column has

That last one becomes very important in Week 6.

## 4. Loading real data

Toy data is fine for learning. Now let's load something real.

Two common ways:

**A. From a CSV file** (you did this manually in Week 2):
```python
df = pd.read_csv("data.csv")
```

**B. From a built-in sample dataset** (needs the `seaborn` library):
```python
import seaborn as sns
df = sns.load_dataset("titanic")
```

We will use the **Titanic** dataset — passenger records from the 1912 Titanic ship.
It is the most famous beginner dataset in data science.

In [32]:
!pip install seaborn

  Using cached seaborn-0.13.2-py3-none-any.whl.metadata (5.4 kB)
  Using cached cycler-0.12.1-py3-none-any.whl.metadata (3.8 kB)
  Using cached pyparsing-3.3.2-py3-none-any.whl.metadata (5.8 kB)
Using cached seaborn-0.13.2-py3-none-any.whl (294 kB)
   ---------------------------------------- 0.0/9.3 MB ? eta -:--:--
   - -------------------------------------- 0.3/9.3 MB ? eta -:--:--
   -- ------------------------------------- 0.5/9.3 MB 1.4 MB/s eta 0:00:07
   -- ------------------------------------- 0.5/9.3 MB 1.4 MB/s eta 0:00:07
   -- ------------------------------------- 0.5/9.3 MB 1.4 MB/s eta 0:00:07
   -- ------------------------------------- 0.5/9.3 MB 1.4 MB/s eta 0:00:07
   --- ------------------------------------ 0.8/9.3 MB 516.0 kB/s eta 0:00:17
   ------ --------------------------------- 1.6/9.3 MB 1.1 MB/s eta 0:00:07
   ------ --------------------------------- 1.6/9.3 MB 1.1 MB/s eta 0:00:07
   ------- -------------------------------- 1.8/9.3 MB 982.7 kB/s eta 0:00:08
 

In [33]:
import seaborn as sns

print(sns.get_dataset_names())



['anagrams', 'anscombe', 'attention', 'brain_networks', 'car_crashes', 'diamonds', 'dots', 'dowjones', 'exercise', 'flights', 'fmri', 'geyser', 'glue', 'healthexp', 'iris', 'mpg', 'penguins', 'planets', 'seaice', 'taxis', 'tips', 'titanic']


In [34]:
titanic = sns.load_dataset("titanic")
print("Loaded! Shape:", titanic.shape)

Loaded! Shape: (891, 15)


In [35]:
titanic

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


In [36]:
titanic.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='str')

In [37]:
titanic.index

RangeIndex(start=0, stop=891, step=1)

In [38]:
titanic.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [39]:
titanic.tail()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
886,0,2,male,27.0,0,0,13.00,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.00,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.45,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.00,C,First,man,True,C,Cherbourg,yes,True
890,0,3,male,32.0,0,0,7.75,Q,Third,man,True,NaN,Queenstown,no,True


In [40]:
titanic.head(10)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
5,0,3,male,NaN,0,0,8.4583,Q,Third,man,True,NaN,Queenstown,no,True
6,0,1,male,54.0,0,0,51.8625,S,First,man,True,E,Southampton,no,True
7,0,3,male,2.0,3,1,21.0750,S,Third,child,False,NaN,Southampton,no,False
8,1,3,female,27.0,0,2,11.1333,S,Third,woman,False,NaN,Southampton,yes,False
9,1,2,female,14.0,1,0,30.0708,C,Second,child,False,NaN,Cherbourg,yes,False


In [41]:
titanic.head(50)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
5,0,3,male,NaN,0,0,8.4583,Q,Third,man,True,NaN,Queenstown,no,True
6,0,1,male,54.0,0,0,51.8625,S,First,man,True,E,Southampton,no,True
7,0,3,male,2.0,3,1,21.0750,S,Third,child,False,NaN,Southampton,no,False
8,1,3,female,27.0,0,2,11.1333,S,Third,woman,False,NaN,Southampton,yes,False
9,1,2,female,14.0,1,0,30.0708,C,Second,child,False,NaN,Cherbourg,yes,False


In [42]:
titanic.tail(10)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
881,0,3,male,33.0,0,0,7.8958,S,Third,man,True,NaN,Southampton,no,True
882,0,3,female,22.0,0,0,10.5167,S,Third,woman,False,NaN,Southampton,no,True
883,0,2,male,28.0,0,0,10.5000,S,Second,man,True,NaN,Southampton,no,True
884,0,3,male,25.0,0,0,7.0500,S,Third,man,True,NaN,Southampton,no,True
885,0,3,female,39.0,0,5,29.1250,Q,Third,woman,False,NaN,Queenstown,no,False
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True
890,0,3,male,32.0,0,0,7.7500,Q,Third,man,True,NaN,Queenstown,no,True


In [43]:
titanic.tail(2)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
889,1,1,male,26.0,0,0,30.00,C,First,man,True,C,Cherbourg,yes,True
890,0,3,male,32.0,0,0,7.75,Q,Third,man,True,NaN,Queenstown,no,True


**Note on `pd.read_csv()` in Colab:**
If you upload your own CSV to Colab, the path is usually just the file name:
```python
df = pd.read_csv("my_file.csv")
```
If the file is in Google Drive, you mount Drive first and then use the full path.
We will use `read_csv` a lot in Week 6.

## 5. The 5-command ritual

Every time you meet a new dataset, run these five commands. Every time. No exceptions.

1. `.head()` — what does it look like?
2. `.shape` — how big is it?
3. `.info()` — what are the columns and types?
4. `.describe()` — what do the numbers look like?
5. `.isna().sum()` — what is missing?

This takes 30 seconds and saves you hours.

In [44]:
# 1. head() — the first 5 rows
titanic.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [45]:
# You can ask for any number of rows
titanic.head(3)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True


In [46]:
# tail() — the last rows
titanic.tail(3)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
888,0,3,female,NaN,1,2,23.45,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.00,C,First,man,True,C,Cherbourg,yes,True
890,0,3,male,32.0,0,0,7.75,Q,Third,man,True,NaN,Queenstown,no,True


In [51]:
# sample() — random rows. Useful because head() can be misleading if data is sorted.
titanic.sample(10)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
432,1,2,female,42.0,1,0,26.0000,S,Second,woman,False,NaN,Southampton,yes,False
868,0,3,male,NaN,0,0,9.5000,S,Third,man,True,NaN,Southampton,no,True
225,0,3,male,22.0,0,0,9.3500,S,Third,man,True,NaN,Southampton,no,True
156,1,3,female,16.0,0,0,7.7333,Q,Third,woman,False,NaN,Queenstown,yes,True
770,0,3,male,24.0,0,0,9.5000,S,Third,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
363,0,3,male,35.0,0,0,7.0500,S,Third,man,True,NaN,Southampton,no,True
556,1,1,female,48.0,1,0,39.6000,C,First,woman,False,A,Cherbourg,yes,False
437,1,2,female,24.0,2,3,18.7500,S,Second,woman,False,NaN,Southampton,yes,False
511,0,3,male,NaN,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [52]:
# 2. shape
print("Rows, Columns:", titanic.shape)
print("Number of passengers:", titanic.shape[0])

Rows, Columns: (891, 15)
Number of passengers: 891


In [53]:
# 3. info()
titanic.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    str     
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    str     
 8   class        891 non-null    category
 9   who          891 non-null    str     
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    str     
 13  alive        891 non-null    str     
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), str(5)
memory usage: 80.7 KB


### Understanding the Titanic columns

| Column | Meaning |
|--------|---------|
| `survived` | 0 = died, 1 = survived |
| `pclass` | Ticket class as a number: 1, 2, 3 |
| `sex` | male / female |
| `age` | Age in years |
| `sibsp` | Number of siblings / spouses aboard |
| `parch` | Number of parents / children aboard |
| `fare` | Ticket price |
| `embarked` | Port they boarded from (C, Q, S) |
| `class` | Ticket class as text: First, Second, Third |
| `who` | man / woman / child |
| `alone` | True if travelling alone |

Some columns repeat the same information in different forms (`pclass` and `class`). That is normal in real data.

In [54]:
# 4. describe() — statistics for every numeric column at once
titanic.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


Read this table row by row:
- `count` — how many values are present
- `mean` — the average
- `std` — standard deviation (spread), from Week 4
- `min` / `max` — the extremes
- `25%` / `50%` / `75%` — the quartiles. `50%` is the median.

Already we can see something: `age` has a count of 714 while other columns have 891.
**177 ages are missing.**

In [55]:
# 5. isna().sum() — count the missing values in each column
titanic.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

This is our first look at missing data. We are only **counting** it today.
**Fixing** missing data is the whole topic of Week 6.

## 6. Selecting columns

Two forms, and the difference matters:

- `df['age']` — one set of square brackets → returns a **Series**
- `df[['age', 'fare']]` — two sets → returns a **DataFrame**

The inner brackets are just a **list of column names**. That is why there are two.

In [57]:
# Single column -> Series
ages = titanic["age"]
print(type(ages))
print(ages.head(10))

<class 'pandas.Series'>
0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: age, dtype: float64


In [61]:
# Multiple columns -> DataFrame
titanic[["age", "fare", "survived"]].head(10)

,age,fare,survived
0,22.0,7.2500,0
1,38.0,71.2833,1
2,26.0,7.9250,1
3,35.0,53.1000,1
4,35.0,8.0500,0
5,NaN,8.4583,0
6,54.0,51.8625,0
7,2.0,21.0750,0
8,27.0,11.1333,1
9,14.0,30.0708,1


In [63]:
# Even one column in a list gives a DataFrame
subset = titanic[["age"]]
print(type(subset))
subset.head()

<class 'pandas.DataFrame'>


,age
0,22.0
1,38.0
2,26.0
3,35.0
4,35.0


## 7. Quick summaries

Once you have a column (a Series), you can summarize it instantly.

In [64]:
print("Average age  :", titanic["age"].mean())
print("Oldest       :", titanic["age"].max())
print("Youngest     :", titanic["age"].min())
print("Median age   :", titanic["age"].median())
print("Total fare   :", titanic["fare"].sum())

Average age  : 29.69911764705882
Oldest       : 80.0
Youngest     : 0.42
Median age   : 28.0
Total fare   : 28693.9493


In [65]:
# The mean of a 0/1 column is the percentage of 1s. A very useful trick.
print("Survival rate:", titanic["survived"].mean())

Survival rate: 0.3838383838383838


That number means about 38% of passengers survived. Because `survived` is 0 or 1,
the average **is** the survival rate. Remember this trick — we use it a lot on Day 3.

In [66]:
# value_counts() — how many of each category?
titanic["sex"].value_counts()

sex
male      577
female    314
Name: count, dtype: int64

In [67]:
titanic

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


In [68]:
titanic['class'].value_counts()

class
Third     491
First     216
Second    184
Name: count, dtype: int64

In [69]:
# As percentages instead of counts
titanic["class"].value_counts(normalize=True)

class
Third     0.551066
First     0.242424
Second    0.206510
Name: proportion, dtype: float64

In [71]:
# unique() and nunique()
print("Unique ports:", titanic["embarked"].unique())
print("How many unique ports:", titanic["embarked"].nunique())
print("Unique classes:", titanic["class"].unique())
print("Unique classes:", titanic["class"].nunique())

Unique ports: <StringArray>
['S', 'C', 'Q', nan]
Length: 4, dtype: str
How many unique ports: 3
Unique classes: ['Third', 'First', 'Second']
Categories (3, str): ['First', 'Second', 'Third']
Unique classes: 3


## Wrap-up — today's toolkit

```python
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")   # load
pd.read_csv("file.csv")            # load from CSV

# Look at it (the 5-command ritual)
df.head()        df.tail()       df.sample(5)
df.shape         df.info()       df.describe()
df.isna().sum()

# Select
df["age"]                # one column  -> Series
df[["age", "fare"]]      # many columns -> DataFrame

# Summarize
df["age"].mean()   .max()   .min()   .median()   .sum()
df["sex"].value_counts()
df["class"].unique()   .nunique()
```

### Next class (Day 2)
We stop *looking* at data and start **asking questions of it**:
`.loc` / `.iloc`, boolean filtering, sorting, and `groupby`.

### Practice
Open **`week5_day1_practice.ipynb`** and complete it before Day 2.
Your assignment is in **`week5_day1_assignment.ipynb`**.

Great work today. See you next class. 🚀